feature/ATC-96

erikquinn opened the pull request:
Initial work on improving the scoring system. Opening for visibility and feedback.

erikquinn committed:
Define game events and update score methods

erikquinn committed:
Update scoring throughout the app

n8rzz commented on src/assets/scripts/aircraft/AircraftController.js:
```diff
@@ -195,15 +196,15 @@ export default class AircraftController {
 
             if (aircraft.isStopped() && aircraft.category === 'arrival') {
-                window.gameController.game.score.windy_landing += aircraft.scoreWind('landed');
+                aircraft.scoreWind('landed');
```
This is beyond the scope of your changes, but why is the `scoreWind` method being used on an `aircraft` instance?  Looking at the method:

``` javascript
    scoreWind(action) {
        let score = 0;
        const components = this.getWind();
        const isWarning = true;

        // TODO: these two if blocks could be done in a single switch statement
        if (components.cross >= 20) {
            window.gameController.game_add_to_score(GAME_EVENTS.EXTREME_CROSSWIND_OPERATION);
            window.uiController.ui_log(`${this.getCallsign()} ${action} with major crosswind'`, isWarning);
        } else if (components.cross >= 10) {
            window.gameController.game_add_to_score(GAME_EVENTS.HIGH_CROSSWIND_OPERATION);
            window.uiController.ui_log(`${this.getCallsign()} ${action} with crosswind'`, isWarning);
        }

        if (components.head <= -10) {
            window.gameController.game_add_to_score(GAME_EVENTS.EXTREME_TAILWIND_OPERATION);
            window.uiController.ui_log(`${this.getCallsign()} ${action} with major tailwind'`, isWarning);
        } else if (components.head <= -1) {
            window.gameController.game_add_to_score(GAME_EVENTS.HIGH_TAILWIND_OPERATION);
            window.uiController.ui_log(`${this.getCallsign()} ${action} with tailwind'`, isWarning);
        }

        return score;
    }
```

I don't see much in that method that the `AircraftInstanceModel` should be concerned about.  

Could you add a TODO above the `scoreWind` method that it needs to be moved/refactored?

erikquinn replied:
yeah, I didn't think it made much sense either. will do.

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -2056,18 +2056,18 @@ export default class Aircraft {
 
                         // Glideslope intercept
-                        if(this.altitude > glideslope_altitude + maxAboveGlideslope) {
+                        if (this.altitude > glideslope_altitude + maxAboveGlideslope) {
```
nice catch! 👍

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -1,13 +1,35 @@
+ * Definitions of point values for given score events
+ * @type {Object}
+ */
+export const GAME_EVENTS = {
```
I _love_ how you've centralized these constants and I especially love how you've added the TODO at the top.  I was about to ask if this is where these should go, and I can see you're already thinking ahead.  Great job!

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -25,6 +47,7 @@ export default class GameController {
         this.game.frequency = 1;
         this.game.time = 0;
         this.game.delta = 0;
+        this.game.events = {};
```
As we go through these files more I would love to update class properties to be initialized correctly.  There are _so_ many properties that are initialized as `null`.  Although that may be correct for classes, it is very incorrect for strings and arrays.

``` javascript
const boolean = false;
const number = -1;
const string = '';
const array = [];
const object = {};
const class = null;
```

By initializing a property as invalid you do two things; you inform other developers about the data type and you immediately have and falsey type value for properties that you can check against.  There's more to it than that, but we can talk more about that some other time if you're curious.

You did good here and I'd like to see this pattern continue. 👍

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -37,27 +60,7 @@ export default class GameController {
-            violation: 0,
-            restrictions: 0
-        };
+        this.game.score = 0;
```
yay refactoring!

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,56 @@ export default class GameController {
+        // Initialize event count
+        _forEach(GAME_EVENTS, (v, k) => {
+            prop.game.events[k] = 0;
+        });
```
be sure to spell these two values out and give them contextual meaning.  what to `v` and `k` mean?

``` javascript
_forEach(GAME_EVENTS, (gameEvent, key) => {
    prop.game.events[key] = 0;
});
```

in this case its rather subtle, but it makes it way easier for other developers to know whats going on without having to dig too far.  Sometimes it is obvious, like `i` in a for loop.  Other times, though, it is not.

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,56 @@ export default class GameController {
+    game_get_weighted_score() {
+        const hoursPlayed = this.game_time() / TIME.ONE_HOUR_IN_SECONDS;
+        const scorePerHour = this.game.score / hoursPlayed;
+        return scorePerHour;
```
be sure to add a space between the consts and return.

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,56 @@ export default class GameController {
-        return score;
+    game_add_to_score(scoreEvent) {
+        prop.game.score[scoreEvent] += 1;
+        prop.game.score.points += SCORE_EVENTS_VALUES[scoreEvent];
```
👍 on abstracting this out to a method and utilizing your constants.  Now it looks like _all_ score updates will run through this method, right?  Thats a good thing!

erikquinn replied:
yep, that's the idea! :smile:

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,56 @@ export default class GameController {
+        // Reset events
+        _forEach(this.game.events, (v, k) => {
+            this.game.events[k] = 0;
+        });
```
same as the other `_forEach` be sure to spell `v` and `k` out and provide contextual meaning to them.

👍 👍  for using `_forEach` with this new stuff, by the way.

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -338,10 +290,8 @@ export default class GameController {
-
-        if (score !== prop.game.last_score) {
-            this.updateScore(score);
+        if (this.game.score !== prop.game.last_score) {
```
should this still be `prop.game` or `this.game`?

erikquinn replied:
This was one of the things I was actually going to ask you, since I see a mix. I know the intent is to change all of these in here to `this` and remove `prop`, reference issue #43.

erikquinn committed:
Align $("#score") right and allow 6 characters of room

erikquinn commented on src/assets/scripts/game/GameController.js:
```diff
@@ -1,13 +1,35 @@
+ * Definitions of point values for given score events
+ * @type {Object}
+ */
+const GAME_EVENTS = {
```
@n8rzz You'll see I have removed the export here, and the imports to other files. After actually using my brain, I realized that importing `GAME_EVENTS`, and trying to say `window.gameController.record_game_event(GAME_EVENTS.ARRIVAL);` does not make any sense because `GAME_EVENTS.ARRIVAL` has a value of `10`, which is of no use when you try to add the event. I've instead just fed `'ARRIVAL'` in each call of `record_game_event()` (formerly `game_add_to_score`. Not sure if there's a better way to do it, but that's where I'm at with it now.

n8rzz replied:
Yeah, that is a problem.  The solution here is actually create two objects: One with a key and a value `ARRIVAL: 10` and one that is simply an enumeration: `ARRIVAL: 'ARRIVAL'`.  That may seem a little goofy, but this allows for slightly increased performance and consolidation of the event names.  Just like you've done with the scores.  You could call them `GAME_EVENT_NAMES` for the enum and `GAME_EVENT_SCORE_MODIFIERS`, or something along those lines.  Then, in the functions that call `record_game_event` you simply reference the enum: 

```
window.gameController.record_game_event('DEPARTURE');
// becomes
window.gameController.record_game_event(GAME_EVENT_NAMES.DEPARTURE);
```

That will help eliminate the possibility of typos, it references the same object and keeps everything in one place, and there is a single source of truth for what each event means.

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,61 @@ export default class GameController {
-        score *= 500;
+    record_game_event(gameEvent) {
+        if (Object.keys(GAME_EVENTS).indexOf(gameEvent) === -1) {
+            throw new TypeError(`Expected a game event listed in GAME_EVENTS, but instead received ${gameEvent}`);
```
👍 

Good use of `throw` here.

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,61 @@ export default class GameController {
-        score = score / (this.game_time() / 60);
-        score *= 500;
+    record_game_event(gameEvent) {
+        if (Object.keys(GAME_EVENTS).indexOf(gameEvent) === -1) {
```
you could use `_has()` here.

`_has(GAME_EVENTS, gameEvent);`

erikquinn replied:
done.

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,61 @@ export default class GameController {
+        // Initialize event count
+        _forEach(GAME_EVENTS, (gameEvent, key) => {
+            this.game.events[key] = 0;
+        });
```
I like what you've done here.  Ordinarily I would lean towards specificity, explicitly set each value, I think it makes sense here because you already have the constant defined.

I would like to see this `_forEach` in it's own method, though.  It can still be called here, but this logic should live within a method.

erikquinn replied:
done.

n8rzz reviewed (commented):
A few more items for you.  Overall this is looking really good, nice work!

If you're feeling up to it, it would be great if you could also update all the `prop.game` instances to be `this.game`.  You don't have to, but since your in the file you might as well?  Be sure to test it afterwards to make sure there aren't any side-effects (shouldn't be).

You're doing very well, keep it up!

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,61 @@ export default class GameController {
-        score -= prop.game.score.violation;
-        score -= prop.game.score.restrictions * 10;
+    game_get_weighted_score() {
+        const hoursPlayed = this.game_time() / TIME.ONE_HOUR_IN_SECONDS;
```
👍 
good use of constants here

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,61 @@ export default class GameController {
-
-        score = score / (this.game_time() / 60);
-        score *= 500;
+    record_game_event(gameEvent) {
```
👍 👍 
For abstracting this to a method _and_ centralizing where game events are updated

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -68,107 +71,61 @@ export default class GameController {
-        prop.game.score.warning = 0;
-        prop.game.score.windy_landing = 0;
-        prop.game.score.windy_takeoff = 0;
+    game_reset_score_and_events() {
```
👍 
For centralizing this method

erikquinn committed:
Replace "prop.game" with "GameController.game"

erikquinn commented:
Updated this reference the feedback, and also replaced all instances of `prop` from `GameController.js` with this, and of course references to `prop.game` to `window.gameController.game`. Does not entirely resolve #43 obviously.

I'm much happier with this PR now; thanks for the help with it. Let me know if you have anything else you want changed. 😄

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -1,13 +1,51 @@
+    SEPARATION_LOSS: -200
+};
+
+const GAME_EVENTS_POINT_VALUES = {
```
did you mean for this to be `GAME_EVENTS`?

n8rzz commented on src/assets/scripts/game/GameController.js:
```diff
@@ -1,13 +1,51 @@
+ * Definitions of point values for given score events
+ * @type {Object}
+ */
+const GAME_EVENTS = {
```
Did you mean for this to be `GAME_EVENT_POINT_VALUE`?

n8rzz commented on src/assets/scripts/InputController.js:
```diff
@@ -447,7 +447,7 @@ export default class InputController {
                 // shortKeys in use
-                if (prop.game.option.get('controlMethod') === 'arrows') {
+                if (window.gameController.game.option.get('controlMethod') === 'arrows') {
                     this.$commandInput.val(`${currentCommandInputValue} \u2BA2`);
```
👍 for moving these to the `gameController`!

Side note: Out of scope for you changes, but, there looks to be _a lot_ of room for enumerations in the `InputController`.  I'll add an issue for that (#123)

n8rzz commented on src/assets/scripts/aircraft/AircraftConflict.js:
```diff
@@ -136,7 +136,7 @@ export default class AircraftConflict {
             window.uiController.ui_log(`${this.aircraft[0].getCallsign()} collided with ${this.aircraft[1].getCallsign()}`, isWarning);
 
-            prop.game.score.hit += 1;
+            window.gameController.events_recordNew('COLLISION');
```
make sure to update these to use the new enumeration.  This should now be `GAME_EVENTS.COLLISION`.

n8rzz commented on src/assets/scripts/aircraft/AircraftController.js:
```diff
@@ -195,15 +195,15 @@ export default class AircraftController {
                 ]);
 
-                window.gameController.game.score.arrival += 1;
+                window.gameController.events_recordNew('ARRIVAL');
```
make sure to update these to use the new enumeration.  This should now be `GAME_EVENTS.ARRIVAL`.

n8rzz commented on src/assets/scripts/aircraft/AircraftInstanceModel.js:
```diff
@@ -464,10 +464,10 @@ export default class Aircraft {
             if (this.isHeadingInsideDepartureWindow()) {
                 this.radioCall('switching to center, good day', 'dep');
-                prop.game.score.departure += 1;
+                window.gameController.events_recordNew('DEPARTURE');
```
make sure to update these to use the new enumeration.  There are several in this file that are still strings.

n8rzz reviewed (commented):
Looking better!  There are several strings that now have an enum, so those should be updated.  But other than that, I like how this is looking.  Centralizing the game stuff to the window method will make it much easier to remove those methods from the window in the near future.

erikquinn commented:
GAHHH i had a bad feeling something got messed up with all the `git stash`ing I was doing to try to reorganize, but somehow figured every little thing was gonna be alrjght... I'll fix it when I get home...

n8rzz commented:
don't worry, about a thing. 😉 

Looks like there is also a conflict with the `AirportModel`.

erikquinn commented:
Yeah I'm sure it's all sorts of messed up... I get off work here in about 2 hours so I'll deal with it a bit after then.

n8rzz commented:
No worries, take your time.  I'm hoping to still be online at that point, as long as this headache goes away. 👊

erikquinn commented:
Finally pushed what was MEANT to be pushed last night!

n8rzz committed:
Merge branch 'develop' of https://github.com/n8rzz/atc into feature/ATC-96

Conflicts:
- src/assets/scripts/airport/AirportModel.js

erikquinn commented:
Phew! :stuck_out_tongue_winking_eye:
